<a href="https://colab.research.google.com/github/gustavocac/FromScratch/blob/master/NLP_101_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1><div align="center">Processamento de Linguagem Natural</div></h1>
<div align="center">Gustavo C. A Corradi</div>
<div align="center"><a href="http://robotreport.me/">robotreport.me</a></div>
<div align="center">@gustavocorradi</div>

<h2><div align="left">Representação de texto</div></h2>

Aqui veremos em alguns detalhes como podemos representar melhor o texto em nosso aplicativo. 
Vamos começar importando os módulos que usaremos

In [0]:
import string
from collections import Counter
from pprint import pprint
import gzip
import matplotlib.pyplot as plt 
import numpy as np
import urllib
import pandas as pd
import numpy as np
import re
import nltk
from bs4 import BeautifulSoup
%matplotlib inline

Agora importaremos o banco de dados de artigos da revista de radiologia brasileira.

In [0]:

def url_pt_br(number):
    url = "http://www.rb.org.br/detalhe_artigo.asp?id="+str(number)+"&idioma=Portugues"
    print(url)
    return url

def extract_text(url):    
    html = urllib.request.urlopen(url).read()
    soup = BeautifulSoup(html)
    for script in soup(["script", "style"]):
        script.extract()    # rip it out
    # baixar o texto
    text = soup.get_text()
    # quebrar em linhas e remover espaço à esquerda e à direita em cada uma delas
    lines = (line.strip() for line in text.splitlines())
    # quebrar parágrafos  de múltiplas linhas em uma frase cada
    chunks = (phrase.strip() for line in lines for phrase in line.split("  "))
    # apagar linhas em branco
    text = '\n'.join(chunk for chunk in chunks if chunk)
    print(text)
    return text

def loop_under(number):
  cols = ['c1', 'c2', 'c3']
  list=[]
  z=0
  z2 = ' '
  for x in range(800, number):
    #print(x)
    index = z
    numero = x
    url = url_pt_br(x)
    z= z+1
    z1 = extract_text(url)
    z2 = z2 +' '+ z1 
    list.append([index, numero, z1])
    df = pd.DataFrame(list, columns=cols)
  return df, z2

In [4]:
print(string.punctuation)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [6]:
#df, texto = loop_under(840)
df, texto = loop_under(820)

http://www.rb.org.br/detalhe_artigo.asp?id=800&idioma=Portugues
Radiologia Brasileira -
Achados de imagem por tomografia computadorizada e ressonância magnética do os styloideum em atleta sintomático
Radiologia Brasileira - Publicação Científica Oficial do Colégio Brasileiro de Radiologia
AMB - Associação Médica Brasileira CNA - Comissão Nacional de Acreditação
Idioma/Language:
Home
CBR
RB History
Editorial Board
Publication Guidelines
Advanced Search
Current Issue
Previous Issues
Paper Submission
Audience
Supplements
Ahead of Print
Contact us
Vol. 43 nº 3 - Maio / Jun.  of 2010
RELATO DE CASO
Print
Page(s) 207 to 209
Achados de imagem por tomografia computadorizada e ressonância magnética do os styloideum em atleta sintomático
Autho(rs): André Francisco Gomes, Vinícius Chissini Paganella, Mateus Chissini Paganella, Caroline Schwartz Henkin, Giana Giostri
PDF Português
PDF English
Texto em Português
English Text
Descritores: Os styloideum, Edema ósseo trabecular, Imagem por ressonância

In [11]:
texto

'  Radiologia Brasileira -\nAchados de imagem por tomografia computadorizada e ressonância magnética do os styloideum em atleta sintomático\nRadiologia Brasileira - Publicação Científica Oficial do Colégio Brasileiro de Radiologia\nAMB - Associação Médica Brasileira CNA - Comissão Nacional de Acreditação\nIdioma/Language:\nHome\nCBR\nRB History\nEditorial Board\nPublication Guidelines\nAdvanced Search\nCurrent Issue\nPrevious Issues\nPaper Submission\nAudience\nSupplements\nAhead of Print\nContact us\nVol. 43\xa0nº\xa03 - Maio / Jun. \xa0of\xa02010\nRELATO DE CASO\nPrint\nPage(s) 207 to 209\nAchados de imagem por tomografia computadorizada e ressonância magnética do os styloideum em atleta sintomático\nAutho(rs): André Francisco Gomes, Vinícius Chissini Paganella, Mateus Chissini Paganella, Caroline Schwartz Henkin, Giana Giostri\nPDF Português\nPDF English\nTexto em Português\nEnglish Text\nDescritores: Os styloideum, Edema ósseo trabecular, Imagem por ressonância magnética\nKeywords:

Escolhemos uma revista de radiologia bem conhecida, e usaremos como seu textos  em nosso exemplo


## Tokenização
O primeiro passo em qualquer análise é tokenizar o texto. O que isto significa é que extrairemos todas as palavras individuais no texto. Por uma questão de simplicidade, assumiremos que nosso texto está bem formado e que nossas palavras são delimitadas por espaços em branco ou caracteres de pontuação.

In [0]:


def extract_words(text):
    #print(text)
    temp = text.split() 
    temp = [w.replace("-", " ") for w in temp]
    #print(temp)
    # Split the text on whitespace
    text_words = []
    table = str.maketrans('', '', string.punctuation)
    text_words = [w.translate(table) for w in temp]
    
    text_words = [w.lower() for w in text_words]
    text_words[:] = [item for item in text_words if item != ' ']
    return text_words



Após esta etapa, agora temos nosso texto representado como um lista de palavras em letras minúsculas

In [93]:
#df['c4'] = df['c3'].apply(lambda x: extract_words(x))
text_words = extract_words(texto)
print(text_words)

['radiologia', 'brasileira', 'achados', 'de', 'imagem', 'por', 'tomografia', 'computadorizada', 'e', 'ressonância', 'magnética', 'do', 'os', 'styloideum', 'em', 'atleta', 'sintomático', 'radiologia', 'brasileira', 'publicação', 'científica', 'oficial', 'do', 'colégio', 'brasileiro', 'de', 'radiologia', 'amb', 'associação', 'médica', 'brasileira', 'cna', 'comissão', 'nacional', 'de', 'acreditação', 'idiomalanguage', 'home', 'cbr', 'rb', 'history', 'editorial', 'board', 'publication', 'guidelines', 'advanced', 'search', 'current', 'issue', 'previous', 'issues', 'paper', 'submission', 'audience', 'supplements', 'ahead', 'of', 'print', 'contact', 'us', 'vol', '43', 'nº', '3', 'maio', '', 'jun', 'of', '2010', 'relato', 'de', 'caso', 'print', 'pages', '207', 'to', '209', 'achados', 'de', 'imagem', 'por', 'tomografia', 'computadorizada', 'e', 'ressonância', 'magnética', 'do', 'os', 'styloideum', 'em', 'atleta', 'sintomático', 'authors', 'andré', 'francisco', 'gomes', 'vinícius', 'chissini', '

Como vimos podemos ser muito mais eficientes representando cada palavra por um número

In [0]:


word_dict = {}
word_list = []
vocabulary_size = 0
text_tokens = []

for word in text_words:
    # If we are seeing this word for the first time, create an id for it and added it to our word dictionary
    if word not in word_dict:
        word_dict[word] = vocabulary_size
        word_list.append(word)
        vocabulary_size += 1
    
    # add the token corresponding to the current word to the tokenized text.
    text_tokens.append(word_dict[word])



Quando estávamos tokenizando nosso texto, também geramos um dicionário word_dict que mapeia palavras para números inteiros e uma word_list que mapeia cada número inteiro para a palavra correspondente.

In [13]:
print("Word list:", word_list, "\n\n Word dictionary:")
print(word_dict)

Word list: ['radiologia', 'brasileira', 'achados', 'de', 'imagem', 'por', 'tomografia', 'computadorizada', 'e', 'ressonância', 'magnética', 'do', 'os', 'styloideum', 'em', 'atleta', 'sintomático', 'publicação', 'científica', 'oficial', 'colégio', 'brasileiro', 'amb', 'associação', 'médica', 'cna', 'comissão', 'nacional', 'acreditação', 'idiomalanguage', 'home', 'cbr', 'rb', 'history', 'editorial', 'board', 'publication', 'guidelines', 'advanced', 'search', 'current', 'issue', 'previous', 'issues', 'paper', 'submission', 'audience', 'supplements', 'ahead', 'of', 'print', 'contact', 'us', 'vol', '43', 'nº', '3', 'maio', '', 'jun', '2010', 'relato', 'caso', 'pages', '207', 'to', '209', 'authors', 'andré', 'francisco', 'gomes', 'vinícius', 'chissini', 'paganella', 'mateus', 'caroline', 'schwartz', 'henkin', 'giana', 'giostri', 'pdf', 'português', 'english', 'texto', 'text', 'descritores', 'edema', 'ósseo', 'trabecular', 'keywords', 'bone', 'marrow', 'magnetic', 'resonance', 'imaging', 'res

Essas duas estruturas de dados já provaram sua utilidade quando convertemos nosso texto em uma lista de tokens.



In [14]:
print(text_tokens)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 0, 1, 17, 18, 19, 11, 20, 21, 3, 0, 22, 23, 24, 1, 25, 26, 27, 3, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 49, 60, 61, 3, 62, 50, 63, 64, 65, 66, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 67, 68, 69, 70, 71, 72, 73, 74, 72, 73, 75, 76, 77, 78, 79, 80, 81, 80, 82, 83, 14, 81, 82, 84, 85, 12, 13, 86, 87, 88, 4, 5, 9, 10, 89, 12, 13, 90, 91, 86, 92, 93, 94, 95, 12, 13, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 23, 111, 12, 13, 8, 112, 103, 113, 96, 114, 115, 116, 117, 118, 119, 12, 120, 121, 97, 62, 3, 12, 13, 122, 86, 87, 88, 103, 123, 5, 9, 10, 14, 15, 124, 125, 122, 126, 127, 128, 129, 130, 131, 12, 13, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 133, 143, 144, 145, 146, 139, 12, 13, 147, 140, 148, 132, 149, 150, 151, 117, 118, 119, 139, 67, 152, 110, 153, 49, 90, 91, 86, 154, 139, 12, 13, 155

Infelizmente, embora essa representação possui algumas limitações graves, principal é o fato de que os computadores assumem naturalmente que os números podem ser operados matematicamente (por adição, subtração, etc.) de um modo que não tenha representação línguistica.

## One-hot encoding
Uma maneira típica de superar essa dificuldade é representar cada palavra por um vetor codificado, onde cada elemento é zero, exceto o que corresponde a uma palavra específica.

In [0]:
def one_hot(word, word_dict):
    """
        Generate a one-hot encoded vector corresponding to *word*
    """
    
    vector = np.zeros(len(word_dict))
    vector[word_dict[word]] = 1
    
    return vector

In [16]:
len(word_dict)

7659

Assim, por exemplo, a palavra "radiologia" seria representada por:



In [17]:
A_hot = one_hot("radiologia", word_dict)
print(A_hot)

array([1., 0., 0., ..., 0., 0., 0.])


Esse vetor tem todo elemento definido como zero, exceto o elemento 0(primeiro elemento da lista), pois:


In [20]:
print(word_dict["brasileira"])
A_hot[1] == 1

1


False

## Bag of words(BOW)

Agora, podemos usar o vetor codificado de cada palavra para produzir uma representação vetorial do texto original, simplesmente adicionando todos os vetores codificados a quente:

In [21]:



text_vector = np.zeros(vocabulary_size)

for word in text_words:
    text_vector[word_dict[word]] += 1
    
print(text_vector)





[120.  62.  18. ...   2.   1.   1.]


In [22]:
text_vector[word_dict["radiologia"]]

120.0

In [23]:
text_vector[word_dict["abdome"]]


16.0

Isso é conhecido como representação do __bag of words__ do texto. Deve-se notar que esses vetores simplesmente contêm o número de vezes que cada palavra aparece em nosso documento.

Uma maneira mais eficiente de produzir o mesmo resultado é usar o módulo __Counter__ padrão:

In [24]:
word_counts = Counter(text_words)
pprint(word_counts)

Streaming output truncated to the last 5000 lines.
         'diameter': 2,
         'utilized': 2,
         'measures': 2,
         'birth': 2,
         'there': 2,
         'means': 2,
         'typical': 2,
         'variations': 2,
         'sadt': 2,
         'hepatomegalia': 2,
         'infância': 2,
         'ultrassonográfico': 2,
         'comprimento': 2,
         'reparos': 2,
         'resultando': 2,
         'precisão': 2,
         'casuística': 2,
         'pais': 2,
         '301': 2,
         '5154': 2,
         '283': 2,
         '4846': 2,
         '023': 2,
         '838': 2,
         '422': 2,
         'usp': 2,
         'liga': 2,
         'puericultura': 2,
         'agep': 2,
         'creche': 2,
         'emei': 2,
         'selecionado': 2,
         '111': 2,
         'pôndero estatural': 2,
         'incluídas': 2,
         'cujos': 2,
         'participação': 2,
         'atendidas': 2,
         'transdutores': 2,
         'multifrequencial': 2,
         'b

A partir do qual podemos gerar facilmente os dados __text_vector__ e __word_dict__


In [0]:
items = list(word_counts.items())

# Extract word dictionary and vector representation
word_dict2 = dict([[items[i][0], i] for i in range(len(items))])
text_vector2 = [items[i][1] for i in range(len(items))]

In [87]:
print("Text vector:", text_vector2, "\n\nWord dictionary:")
pprint(word_dict2)

Streaming output truncated to the last 5000 lines.
 'dorsi': 6862,
 'dorso': 218,
 'dos': 237,
 'dosagem': 1185,
 'dosaram': 4599,
 'dose': 4330,
 'dose area': 4360,
 'dose área': 4329,
 'doses': 4302,
 'dosimetria': 4321,
 'dosimetric': 4895,
 'dosimetry': 4323,
 'dosimétrico': 6368,
 'douro': 6428,
 'doutor': 3859,
 'downar': 5823,
 'dp': 2245,
 'dr': 576,
 'dra': 3856,
 'draisma': 3048,
 'drenagem': 4029,
 'droga': 2514,
 'drogas': 474,
 'du': 3748,
 'duarte': 7392,
 'duas': 960,
 'duckworth': 6855,
 'due': 3055,
 'dukes': 1039,
 'duodenal': 3994,
 'dupla': 1048,
 'duplication': 7113,
 'duplications': 7356,
 'duplicação': 7089,
 'dura': 242,
 'durante': 335,
 'duração': 5430,
 'during': 4403,
 'dutra': 7650,
 'dv': 4919,
 'dw': 6849,
 'dynamic': 6814,
 'dysfunction': 5371,
 'dão': 7175,
 'década': 1327,
 'déficit': 3601,
 'dúvida': 4830,
 'e': 8,
 'e mail': 585,
 'e4 7': 6967,
 'ea': 2233,
 'each': 3247,
 'ear': 747,
 'earlier': 7581,
 'early': 1743,
 'eberhardt': 3042,
 'ecg': 1841


### **Term Frequency**


A representação vetorial do conjunto de palavras introduzida acima depende simplesmente da frequência de ocorrência de cada palavra. Seguindo uma longa tradição de dar nomes sofisticados a idéias simples, isso é conhecido como Term Frequency.

Intuitivamente, esperamos que a frequência com que uma determinada palavra seja mencionada corresponda à relevância dessa palavra para o trecho de texto que estamos considerando. 

In [0]:
data=texto

In [103]:
counts = Counter(data)
sorted_counts = sorted(items, key=lambda x:x[1], reverse=True)
for word, count in sorted_counts[:10]:
    print(word, count)



de 1818
e 1079
a 1078
do 689
o 641
em 583
 571
da 500
com 417
que 352


Surpreendentemente, descobrimos que as palavras mais comuns não são particularmente significativas. De fato, essa é uma ocorrência comum no Processamento de linguagem natural. As palavras mais frequentes são normalmente auxiliares devido às regras gramaticais.
Por outro lado, também há um grande número de palavras que ocorrem com pouca frequência, como pode ser facilmente visto olhando a distribuição de frequência da palavra.

## Stopwords

Uma técnica comum para simplificar as tarefas da NLP é remover o que é conhecido como Stopwords, palavras muito frequentes, mas sem sentido. Se simplesmente removermos as 100 palavras mais comuns, reduziremos significativamente a quantidade de dados que devemos considerar e perderemos pouca informação.

In [67]:

# Extract word dictionary and vector representation
word_dict2 = dict([[items[i][0], i] for i in range(len(items))])
text_vector2 = [items[i][1] for i in range(len(items))]
counts = Counter(data)
sorted_counts = sorted(items, key=lambda x:x[1], reverse=True)
for word, count in sorted_counts[:10]:
    print(word, count)

de 1818
e 1079
a 1078
do 689
o 641
em 583
 571
da 500
com 417
que 352


In [111]:
stopwords = set([word for word, count in sorted_counts[:100]])

clean_data = []
text_words2 = extract_words(texto)
for word in text_words2:
    if word not in stopwords:
      #print(word)
      clean_data.append(word)

print("Original size:", len(data))
print("Clean size:", len(clean_data))
print("Reduction:", 1-len(clean_data)/len(data))

Original size: 268379
Clean size: 23784
Reduction: 0.9113790572287698


Na prática, não removemos simplesmente as palavras mais comuns em nosso corpus, mas uma lista com curadoria manual de palavras-chave. Listas para dezenas de idiomas e aplicativos podem ser facilmente encontradas online.

## Term Frequency/Inverse Document Frequency

Uma maneira de determinar a importância relativa de uma palavra é ver com que frequência ela aparece em vários documentos. Palavras que são relevantes para um tópico específico têm maior probabilidade de aparecer em documentos sobre esse tópico e muito menos em documentos sobre outros tópicos. Por outro lado, palavras menos significativas (como as) serão comuns nos documentos sobre qualquer assunto.

Para medir a frequência do documento de uma palavra, precisamos ter vários documentos. Por uma questão de simplicidade, trataremos cada frase da nossa base de dados como um documento individual:



In [0]:
corpus_text = texto.split('.')
corpus_words = []
for document in corpus_text:
    doc_words = extract_words(document)
    corpus_words.append(doc_words)

Agora, nosso corpus é representado como uma lista de listas de palavras, onde cada lista é apenas a representação da palavra da sentença correspondente.

In [109]:
pprint(corpus_words)

Streaming output truncated to the last 5000 lines.
  'artroscopia',
  'ou',
  'por',
  'cirurgia',
  'aberta',
  'dependendo',
  'do',
  'caso71026 28'],
 ['ressalto',
  'no',
  'tornozelo',
  'o',
  'ressalto',
  'no',
  'tornozelo',
  'é',
  'raro',
  'e',
  'pode',
  'estar',
  'associado',
  'a',
  'subluxação',
  'ou',
  'luxação',
  'de',
  'tendões',
  'e',
  'mais',
  'raramente',
  'é',
  'relatado',
  'associado',
  'com',
  'corpos',
  'intra articulares'],
 ['dentre',
  'os',
  'relatos',
  'na',
  'literatura',
  'a',
  'situação',
  'mais',
  'encontrada',
  'é',
  'a',
  'subluxação',
  'ou',
  'luxação',
  'dos',
  'tendões',
  'fibulares',
  'com',
  'ou',
  'sem',
  'ruptura',
  'tendínea',
  'concomitante29'],
 ['na',
  'maioria',
  'dos',
  'casos',
  'registrados',
  'se',
  'encontra',
  'esta',
  'anormalidade',
  'em',
  'atletas',
  'ou',
  'dançarinos'],
 ['a',
  'literatura',
  'cita',
  'como',
  'fator',
  'predisponente',
  'para',
  'a',
  'luxação',
  'd


Vamos agora calcular o número de documentos em que cada palavra aparece:


In [110]:
document_count = {}

for document in corpus_words:
    word_set = set(document)
    
    for word in word_set:
        document_count[word] = document_count.get(word, 0) + 1

pprint(document_count)

Streaming output truncated to the last 5000 lines.
 'dosaram': 1,
 'dose': 36,
 'dose area': 2,
 'dose área': 3,
 'doses': 22,
 'dosimetria': 2,
 'dosimetric': 1,
 'dosimetry': 2,
 'dosimétrico': 1,
 'douro': 2,
 'doutor': 2,
 'downar': 1,
 'dp': 2,
 'dr': 10,
 'dra': 2,
 'draisma': 1,
 'drenagem': 2,
 'droga': 1,
 'drogas': 1,
 'du': 1,
 'duarte': 1,
 'duas': 7,
 'duckworth': 1,
 'due': 2,
 'dukes': 1,
 'duodenal': 4,
 'dupla': 2,
 'duplication': 1,
 'duplications': 1,
 'duplicação': 4,
 'dura': 1,
 'durante': 41,
 'duração': 1,
 'during': 8,
 'dutra': 1,
 'dv': 1,
 'dw': 1,
 'dynamic': 3,
 'dysfunction': 1,
 'dão': 1,
 'década': 6,
 'déficit': 1,
 'dúvida': 1,
 'e': 709,
 'e mail': 12,
 'e4 7': 1,
 'ea': 1,
 'each': 4,
 'ear': 2,
 'earlier': 1,
 'early': 2,
 'eberhardt': 2,
 'ecg': 2,
 'ecocardiograma': 1,
 'ecoou': 1,
 'ectopia': 1,
 'ed': 15,
 'edelson': 1,
 'edema': 6,
 'editor': 6,
 'editora': 1,
 'editores': 1,
 'editorial': 21,
 'editors': 7,
 'edição': 3,
 'edson': 1,
 'edta 5

In [0]:
def inv_doc_freq(corpus_words):
    number_docs = len(corpus_words)
    
    document_count = {}

    for document in corpus_words:
        word_set = set(document)

        for word in word_set:
            document_count[word] = document_count.get(word, 0) + 1
    
    IDF = {}
    
    for word in document_count:
        IDF[word] = np.log(number_docs/document_count[word])
        
    
    return IDF

Onde seguimos a convenção de usar o logaritmo da frequência inversa do documento. Isso tem a vantagem numérica de evitar ter que lidar com pequenos números fracionários.

Podemos ver facilmente que o IDF atribui um peso menor às palavras mais comuns e um peso maior às menos frequentes:

In [114]:
IDF = inv_doc_freq(corpus_words)

pprint(IDF)

Streaming output truncated to the last 5000 lines.
 'dosaram': 7.870547844507712,
 'dose': 4.300304276436043,
 'dose area': 7.177782416195197,
 'dose área': 6.7726989146552965,
 'doses': 4.787491742782046,
 'dosimetria': 7.177782416195197,
 'dosimetric': 7.870547844507712,
 'dosimetry': 7.177782416195197,
 'dosimétrico': 7.870547844507712,
 'douro': 7.177782416195197,
 'doutor': 7.177782416195197,
 'downar': 7.870547844507712,
 'dp': 7.177782416195197,
 'dr': 5.571393286610456,
 'dra': 7.177782416195197,
 'draisma': 7.870547844507712,
 'drenagem': 7.177782416195197,
 'droga': 7.870547844507712,
 'drogas': 7.870547844507712,
 'du': 7.870547844507712,
 'duarte': 7.870547844507712,
 'duas': 5.926926025970411,
 'duckworth': 7.870547844507712,
 'due': 7.177782416195197,
 'dukes': 7.870547844507712,
 'duodenal': 6.4853983032037545,
 'dupla': 7.177782416195197,
 'duplication': 7.870547844507712,
 'duplications': 7.870547844507712,
 'duplicação': 6.4853983032037545,
 'dura': 7.870547844507712,

Você pode considerar usar isso como uma maneira de identificar e remover implicitamente palavras irrelevantes. Caso deseje manter as palavras que aparecem em todos os documentos, basta adicionar um 1. ao argumento do logaritmo acima:
Quando multiplicamos o termo frequência de cada palavra pela frequência inversa do documento, temos uma boa maneira de quantificar a relevância de uma palavra para entender o significado de um documento específico.

In [0]:
def tf_idf(corpus_words):
    IDF = inv_doc_freq(corpus_words)
    
    TFIDF = []
    
    for document in corpus_words:
        TFIDF.append(Counter(document))
    
    for document in TFIDF:
        for word in document:
            document[word] = document[word]*IDF[word]
            
    return TFIDF

In [116]:
tf_idf(corpus_words)

[Counter({'achados': 4.986646137726233,
          'acreditação': 4.743418696367982,
          'advanced': 4.882044059723269,
          'ahead': 4.833632898779901,
          'amb': 4.882044059723269,
          'associação': 4.356708826689592,
          'atleta': 6.7726989146552965,
          'audience': 4.882044059723269,
          'board': 4.882044059723269,
          'brasileira': 14.362475228346138,
          'brasileiro': 4.196449785159254,
          'cbr': 4.882044059723269,
          'científica': 4.743418696367982,
          'cna': 4.882044059723269,
          'colégio': 4.196449785159254,
          'comissão': 4.743418696367982,
          'computadorizada': 4.833632898779901,
          'contact': 4.882044059723269,
          'current': 4.787491742782046,
          'de': 4.228979730228307,
          'do': 3.786898927307612,
          'e': 1.5459707494863493,
          'editorial': 4.833632898779901,
          'em': 1.9705370175872805,
          'guidelines': 4.787491742782046,
  

Agora, temos uma representação vetorial de cada um de nossos documentos que leva em consideração as contribuições informacionais de cada palavra. Cada um desses vetores nos fornece uma representação única de cada documento em que ocorre, possibilitando definir a semelhança de dois documentos, etc.

## Porter Stemmer

Ainda existe, no entanto, um problema com nossa abordagem para representar texto. Como tratamos cada palavra como um token exclusivo e completamente independente de todas as outras, para documentos grandes, acabaremos com muitas variações da mesma palavra, como conjugações verbais, advérbios e substantivos correspondentes, etc.

Uma maneira de contornar essa dificuldade é usar o algoritmo stemming para reduzir as palavras à sua versão raiz (ou tronco). O algoritmo Stemming mais famoso é conhecido como ** Porter Stemmer ** e foi introduzido por Martin Porter em 1980 [Programa 14, 130 (1980)] (https://dl.acm.org/citation.cfm?id=275705 )

O algoritmo começa definindo consoantes (C) e vogais (V):

In [0]:
V = set('aeiouy')
C = set('bcdfghjklmnpqrstvwxz')

A raiz de uma palavra é o que resta dessa palavra após a remoção de um final específico. Uma função para fazer isso é fácil de implementar:



In [0]:
def get_stem(suffix, word):
    """
        Extract the stem of a word
    """
    
    if word.lower().endswith(suffix.lower()): # Case insensitive comparison
        return word[:-len(suffix)]

    return None

Isto também define palavras como sequências de vogais e consoantes da forma:

\begin{equation}
[C](VC)^m[V]
\end{equation}

onde $ m $ é chamado de **medida** da palavra e [] representa seções opcionais.

In [0]:
def measure(orig_word):
    """
        Calculate the "measure" m of a word or stem, according to the Porter Stemmer algorthim
    """
    
    word = orig_word.lower()

    optV = False
    optC = False
    VC = False
    m = 0

    pos = 0

    # We can think of this implementation as a simple finite state machine
    # looks for sequences of vowels or consonants depending of the state
    # in which it's in, while keeping track of how many VC sequences it
    # has encountered.
    # The presence of the optional V and C portions is recorded in the
    # optV and optC booleans.
    
    # We're at the initial state.
    # gobble up all the optional consonants at the beginning of the word
    while pos < len(word) and word[pos] in C:
        pos += 1
        optC = True

    while pos < len(word):
        # Now we know that the next state must be a vowel
        while pos < len(word) and word[pos] in V:
            pos += 1
            optV = True

        # Followd by a consonant
        while pos < len(word) and word[pos] in C:
            pos += 1
            optV = False
        
        # If a consonant was found, then we matched VC
        # so we should increment m by one. Otherwise, 
        # optV remained true and we simply had a dangling
        # V sequence.
        if not optV:
            m += 1

    return m

In [120]:
word = "tomografia"
print(measure(word))

3


O algoritmo Porter aplica sequencialmente uma série de regras de transformação em uma série de 5 etapas (a etapa 1 é dividida em 3 subetapas e a etapa 5 em 2). As regras são aplicadas apenas se uma determinada condição for verdadeira.

Além de especificar possivelmente um requisito na medida de uma palavra, as condições também podem fazer uso de diferentes funções booleanas:

In [0]:
def ends_with(char, stem):
    """
        Checks the ending of the word
    """
    return stem[-1] == char

def double_consonant(stem):
    """
        Checks the ending of a word for a double consonant
    """
    if len(stem) < 2:
        return False

    if stem[-1] in C and stem[-2] == stem[-1]:
        return True

    return False

def contains_vowel(stem):
    """
        Checks if a word contains a vowel or not
    """
    return len(set(stem) & V) > 0 

In [0]:
def apply_rule(condition, suffix, replacement, word):
    """
        Apply Porter Stemmer rule.
        if "condition" is True replace "suffix" by "replacement" in "word"
    """
    
    stem = get_stem(suffix, word)

    if stem is not None and condition is True:
        # Remove the suffix
        word = stem

        # Add the replacement suffix, if any
        if replacement is not None:
            word += replacement

    return word

In [125]:
word = "tomografia"
suffix = "afia"
stem = get_stem(suffix, word)
apply_rule(contains_vowel(stem), suffix, None, word)

'tomogr'

http://snowball.tartarus.org/algorithms/portuguese/stemmer.html

No total, o algoritmo Porter Stemmer aplica várias dezenas de regras (consulte o algoritmo Porter Stemmer (para o idioma inglês) aplica várias dezenas de regras (consulte https://tartarus.org/martin/PorterStemmer/def.txt para obter uma lista completa). A implementação de todos eles é tediosa e propensa a erros, por isso nos abstivemos de fornecer uma implementação completa do algoritmo aqui. Implementações de alta qualidade podem ser encontradas em todas as principais bibliotecas, como NLTK.
As dificuldades de definir regras de correspondência com texto arbitrário não podem ser totalmente resolvidas sem o uso de Expressões Regulares, como nossa implementação de medida acima, um tópico mais avançado que está além do escopo deste aula.